engine/audio/private/audio_system.cpp

```diff
@@ -0,0 +1,159 @@
+#include "components/transform_component.hpp"
+#include "components/transform_helpers.hpp"
+#include "ecs_module.hpp"
+#include "fmod_debug.hpp"
+
+#include <fmod.h>
+#include <ranges>
+
+#include <tracy/Tracy.hpp>
+
+AudioSystem::AudioSystem(ECSModule& ecs, AudioModule& audioModule)
+    : _ecs(ecs)
+    , _audioModule(audioModule)
+{
+}
+void AudioSystem::Update(ECSModule& ecs, MAYBE_UNUSED float dt)
+{
+    ZoneScoped;
+    const auto& listenerView = ecs.GetRegistry().view<AudioListenerComponent>();
+
+    if (!listenerView.empty())
+    {
+        const auto entity = listenerView.front();
+
+        glm::vec3 position {};
+        glm::vec3 velocity {};
+        glm::vec3 forward {};
+        glm::vec3 up {};
+
+        if (ecs.GetRegistry().all_of<TransformComponent>(entity))
+        {
+            position = TransformHelpers::GetWorldPosition(ecs.GetRegistry(), entity);
+
+            auto rotation = TransformHelpers::GetWorldRotation(ecs.GetRegistry(), entity);
+
+            forward = glm::normalize(rotation * glm::vec3(0.0f, -1.0f, 0.0f));
+            up = glm::normalize(rotation * glm::vec3(0.0f, 0.0f, 1.0f));
+        }
+
+        if (ecs.GetRegistry().all_of<RigidbodyComponent>(entity))
+        {
+            const auto& body = ecs.GetRegistry().get<RigidbodyComponent>(entity);
+            velocity = ToGLMVec3(_audioModule._physics->bodyInterface->GetLinearVelocity(body.bodyID));
+        }
+
+        _audioModule.SetListener3DAttributes(position, velocity, forward, up);
+    }
+
+    const auto& emitterView = ecs.GetRegistry().view<AudioEmitterComponent>();
+    for (const auto entity : emitterView)
+    {
+        AudioEmitterComponent& emitter = ecs.GetRegistry().get<AudioEmitterComponent>(entity);
+
+        // Remove sounds and events from the emitter if they are no longer playing
+        std::erase_if(emitter._soundIds, [&](const auto& id)
+            { return !_audioModule.IsSoundPlaying(id); });
+        std::erase_if(emitter._eventIds, [&](const auto& id)
+            { return !_audioModule.IsEventPlaying(id); });
+
+        if (!ecs.GetRegistry().all_of<TransformComponent>(entity))
+        {
+            return;
+        }
+
+        // Get 3d attributes of emitter
+        const glm::vec3 position = TransformHelpers::GetWorldPosition(ecs.GetRegistry(), entity);
+        glm::vec3 velocity = glm::vec3(0.f, 0.f, 0.f);
+
+        const glm::quat rotation = TransformHelpers::GetWorldRotation(ecs.GetRegistry(), entity);
+        const glm::vec3 forward = glm::normalize(rotation * glm::vec3(0.f, -1.f, 0.f));
+        const glm::vec3 up = glm::normalize(rotation * glm::vec3(0.f, 0.f, 1.f));
+
+        if (RigidbodyComponent* rigidBody = ecs.GetRegistry().try_get<RigidbodyComponent>(entity))
+        {
+            velocity = ToGLMVec3(_audioModule._physics->bodyInterface->GetLinearVelocity(rigidBody->bodyID));
+        }
+        // Update 3D position of sounds
+        for (auto soundInstance : emitter._soundIds)
```

Comment:
Why are we handling both events and sounds at the same time? Shouldn't this logic only work with events that then propagate to the sounds?

Reply:
In the case of fmod, sounds and events are 2 different audio types.
It doesn't have anything to do with an event system.
Sounds are simple samples loaded from .mp3 .ogg .wav files.
While FMOD events are their own more powerful audio type including many variables and possible special objects.

Reply:
Ok I got it!